lib/membrane_audio_mixer.ex

```diff
@@ -97,345 +98,238 @@ defmodule Membrane.AudioMixer do
     ]
 
   @impl true
-  def handle_init(_ctx, %__MODULE__{stream_format: stream_format} = options) do
-    if options.native_mixer && !options.prevent_clipping do
-      raise("Invalid element options, for native mixer only clipping preventing one is available")
-    else
-      state =
-        options
-        |> Map.from_struct()
-        |> Map.put(:pads, %{})
-        |> Map.put(:mixer_state, initialize_mixer_state(stream_format, options))
-        |> Map.put(:last_ts_sent, 0)
-
-      {[], state}
+  def handle_init(_ctx, %__MODULE__{} = options) do
+    if options.native_mixer and not options.prevent_clipping do
+      raise "Invalid element options, for native mixer only clipping preventing one is available"
     end
-  end
-
-  @impl true
-  def handle_pad_added(pad, context, state) do
-    offset = context.pads[pad].options.offset
-
-    if offset < 0,
-      do:
-        raise(
-          "Wrong offset value: #{offset}, audio mixer only allows offset value to be non negative."
-        )
 
     state =
-      put_in(
-        state,
-        [:pads, pad],
-        %{queue: <<>>, ready_to_mix?: false}
-      )
+      options
+      |> Map.from_struct()
+      |> Map.merge(%{pads_data: %{}, last_ts_sent: 0, sample_size: nil, mixer_state: nil})
 
     {[], state}
   end
 
   @impl true
-  def handle_pad_removed(pad, _context, state) do
-    state = Bunch.Access.delete_in(state, [:pads, pad])
+  def handle_pad_added(pad, ctx, state) do
+    offset = ctx.pad_options.offset
 
-    {[], state}
-  end
+    if offset < 0 do
+      raise "Wrong offset value: #{offset}, audio mixer only allows offset value to be non negative."
+    end
 
-  @impl true
-  def handle_playing(_context, %{stream_format: %RawAudio{} = stream_format} = state) do
-    {[stream_format: {:output, stream_format}], state}
-  end
+    pad_data = %{queue: <<>>, synchronized?: false, offset: offset}
+    state = put_in(state, [:pads_data, pad], pad_data)
 
-  def handle_playing(_context, %{stream_format: nil} = state) do
     {[], state}
   end
 
   @impl true
-  def handle_demand(:output, size, :bytes, context, state) do
-    frame_size = RawAudio.frame_size(context.pads.output.stream_format)
-    do_handle_demand(size + frame_size, state)
+  def handle_pad_removed(pad, _ctx, state) do
+    state = Bunch.Access.delete_in(state, [:pads_data, pad])
+    {[], state}
   end
 
   @impl true
-  def handle_demand(:output, _buffers_count, :buffers, _context, %{stream_format: nil} = state) do
+  def handle_demand(:output, _size, _demand, ctx, state)
+      when ctx.pads.output.stream_format == nil do
```

Comment:
I think it would be enough to check if we have non-nil `stream_format` in state (even if we receive it via options, we could make a demand even when the `RemoteStream` is not yet received on the :input pad)

Reply:
We cannot do it because before `handle_stream_format` field `sample_size` is unset and we need it to make a demand